private/buf/buflsp/file.go

```diff
@@ -0,0 +1,675 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file defines file manipulation operations.
+
+package buflsp
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"slices"
+	"strings"
+
+	"github.com/bufbuild/buf/private/buf/bufworkspace"
+	"github.com/bufbuild/buf/private/bufpkg/bufanalysis"
+	"github.com/bufbuild/buf/private/bufpkg/bufcheck"
+	"github.com/bufbuild/buf/private/bufpkg/bufimage"
+	"github.com/bufbuild/buf/private/bufpkg/bufmodule"
+	"github.com/bufbuild/buf/private/pkg/ioext"
+	"github.com/bufbuild/buf/private/pkg/tracing"
+	"github.com/bufbuild/protocompile"
+	"github.com/bufbuild/protocompile/ast"
+	"github.com/bufbuild/protocompile/linker"
+	"github.com/bufbuild/protocompile/parser"
+	"github.com/bufbuild/protocompile/protoutil"
+	"github.com/bufbuild/protocompile/reporter"
+	"github.com/gofrs/uuid/v5"
+	"go.lsp.dev/protocol"
+	"go.opentelemetry.io/otel/attribute"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+const descriptorPath = "google/protobuf/descriptor.proto"
+
+// file is a file that has been opened by the client.
+//
+// Mutating a file is thread-safe.
+type file struct {
+	// lsp and uri are not protected by file.lock; they are immutable after
+	// file creation!
+	lsp *lsp
+	uri protocol.URI
+
+	// All variables after this lock variables are protected by file.lock.
+	//
+	// NOTE: this package must NEVER attempt to acquire a lock on a file while
+	// holding a lock on another file. This guarantees that any concurrent operations
+	// on distinct files can always make forward progress, even if the information they
+	// have is incomplete. This trades off up-to-date accuracy for responsiveness.
+	//
+	// For example, suppose g1 locks a.proto, and then attempts to lock b.proto
+	// because it followed a pointer in importMap. However, in the meantime, g2
+	// has acquired b.proto's lock already, and attempts to acquire a lock to a.proto,
+	// again because of a pointer in importMap. This will deadlock, and it will
+	// deadlock in such a way that will be undetectable to the Go scheduler, so the
+	// LSP will hang forever.
+	//
+	// This seems like a contrived scenario, but it can happen if a user creates two
+	// mutually-recursive Protobuf files. Although this is not permitted by Protobuf,
+	// the LSP must handle this invalid state gracefully.
+	//
+	// This is enforced by mutex.go.
+	lock mutex
+
+	text    string
+	version int32
```

Comment:
We might want to document what `version` here refers to and how it should be used.
And maybe `%s/text/content`, but I'm low confidence on this one (it makes sense to me either way, since this is the `string` representation of content).